Add CRUD options for gopass

andyfase opened the pull request:
## Description

Extends the gopass extension to have full CRUD operations. This PR adds the CUD (Create, Update and Delete) options.

Screencast shows the use of the new features.

## Screencast


https://github.com/user-attachments/assets/8dfee5aa-51e5-4e12-bd3d-d32a749f91b7



## Checklist

- [ x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [ x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [ x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [ x] I checked that files in the `assets` folder are used by the extension itself
- [ x] I checked that assets used by the `README` are placed outside of the `metadata` folder

andyfase commented:
Note. Also pushed support to sync gopass stores. This is needed to ensure password changes made on other machines and pushed to git are synchronized with the local password store.

pernielsentikaer commented:
Thanks for the contribution here 🔥

@bake do you want to check this?

bake commented on extensions/gopass/src/create-edit.tsx:
```diff
@@ -0,0 +1,226 @@
+
+  const getInputPassword = async () => {
+    if (!inputPassword) return;
+    const password = await gopass.showAll(inputPassword);
```
We could simply use `gopass.password(inputPassword)` and `gopass.show(inputPassword).then((details) => details.join("\n"))`. I'm not sure about this, there's a comment in `gopass.ts` about this :)

bake commented on extensions/gopass/src/gopass.ts:
```diff
@@ -47,10 +74,34 @@ async function otp(entry: string): Promise<string> {
+  password: string;
+  attributes: string[];
+}
+async function showAll(entry: string): Promise<showAllResponse> {
```
I’m not sure if we really need this function. We should either remove `showAll()` and replace it with calls to `password()` and `show()` or remove `password()` and rewrite `show()` to return password and details similar to this function. I'm not sure which way I'd prefer. What do you think?

andyfase replied:
So I went with rewrite the old `show`. initially I added a new method to avoid large re-factors  but i didnt check how often that method is actually called - and its only once :-)

I wanted to avoid the multi-call option as there is no need to call the CLI twice imo

bake replied:
That's a good point. I don't think it would have be a problem to call the CLI twice but your solution is way more clean, thank you :)

bake commented on extensions/gopass/src/gopass.ts:
```diff
@@ -39,6 +43,29 @@ async function list({ limit = -1, prefix = "", directoriesFirst = false, stripPr
+  numbers: boolean
+): Promise<string[]> {
+  if (type === "xkcd") {
+    return await gopass([
```
The `await` here and in line 64 is unnecessary. This is a very minor nitpick 😅

andyfase replied:
done

bake reviewed (changes requested):
Awesome PR, thank you so much! I wanted an option to create passwords for a while! 🎉

andyfase commented on extensions/gopass/src/create-edit.tsx:
```diff
@@ -0,0 +1,226 @@
+        </>
+      )}
+
+      <Form.TextField
```
So i implemented the checkbox method - I definitely want the ability to self-alter the suggested password (hence going with clear-text in the first place) but get what you mean about the security aspect

The code to support both (with a checkbox) is a bit kludgy but it works :-) - It means i have to ignore the actual field values in validation and updates. I commented appropriately so its fairly obvious

bake replied:
I really like the user experience, thank you!

andyfase commented on extensions/gopass/src/create-edit.tsx:
```diff
@@ -0,0 +1,226 @@
+  const { handleSubmit, itemProps } = useForm<SignUpFormValues>({
+    async onSubmit(values: SignUpFormValues) {
+      const inputString = `${values.password}\n${values.additionalAttributes}\n`;
+      console.log(inputString);
```
removed

andyfase commented on extensions/gopass/src/create-edit.tsx:
```diff
@@ -0,0 +1,226 @@
+      },
+      additionalAttributes: (value: string | undefined) => {
+        if (value !== undefined && value !== "") {
+          const rAttributes = /^(?:(?:[^:,\n]+:[^:,\n]+)\n)*(?:[^:,\n]+:[^:,\n]+)$/;
```
changed the regex to support it but i remain unsure if `foo:bar:baz` is actually valid? I guess any value is valid.

andyfase commented:
@bake thanks for the feedback - I implemented the changes you requested / issues you pointed out. Also I added

- Check for existing password existence  - on new password
- Check for password name change  - on edit password (requires a save and a move operation)
- Fixed the fact that the password would fail to generate 1 times out of 10

Let me know what you think.

bake commented:
Thank you again for your additions. I'm really looking forward to using it as soon as the PR is merged :)

andyfase commented:
thanks @bake - just FYI I pushed some update to the create-edit view to make the re-renders based on updates to the form elements more reliable (moving away from `useEffect` )

@pernielsentikaer this should be good to merge now

pernielsentikaer reviewed (approved):
Hi 👋 

Looks good to me, approved 🔥
